### DISCUSSION TITLE
HELP HELP reading mf6 budget file  for a transient simulation returning error of the invalid imeth code 1085209134

### DISCUSSION THREAD

**Discussion by StellaLiu-gogogo** [modflowpy/flopy#2105] [Q&A] [answered]
Recently I'm encountering a really weird situation. 
I'm working in a linux based HPC envinronment with python 3.9, flopy 3.5version.

The model runs fine in steady state and the .cbb file can be read and analyzed without any problem. 
When I turned the simulation as transient. the model terminates normally and returns reasonable groundwater level results.
in the list file, I could also see the water budget values for each flow component. it also looks reasonable in my case. 

HOWEVER, when I try to use the flopy.util.binaryfile to read my .cbb file, It returns me the following error. 
To my level of understanding, it means the header of the binary file returns a imeth value of 1085209134, but it should be a value of 1-6, representing 6 different methods of storing the water budget records. 

I have no clue why the cbb file I have is with such a weird long number of imeth code. Especially all the rest of the model results just looks perfectly fine and it only happens in this specific case. Since I also used this same set of script for other case studies and it never returns me this weird error. 

gonna be really appreciate if anyone could give some thoughts on how can I fix it.    

`---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
Cell In[28], line 3
      1 import flopy.utils.binaryfile as bf
      2 path = 'MYPATH/workspace_2/mf6_2l_ss_rch2.cbb'
----> 3 bud = bf.CellBudgetFile(path)
      4 hds = bf.HeadFile('MYPATH/workspace_2/mf6_2l_ss_rch2.hds')
      5 print(type(hds))

File ~/miniconda3/envs/py39/lib/python3.9/site-packages/flopy/utils/binaryfile.py:1037, in CellBudgetFile.__init__(self, filename, precision, verbose, **kwargs)
   1034     raise Exception(f"LayerFile error: unrecognized kwargs: {args}")
   1036 if precision == "auto":
-> 1037     success = self._set_precision("single")
   1038     if not success:
   1039         success = self._set_precision("double")

File ~/miniconda3/envs/py39/lib/python3.9/site-packages/flopy/utils/binaryfile.py:1133, in CellBudgetFile._set_precision(self, precision)
   1130 self.header_dtype = np.dtype(hdt)
   1132 try:
-> 1133     self._build_index()
   1134 except BudgetIndexError:
   1135     success = False

File ~/miniconda3/envs/py39/lib/python3.9/site-packages/flopy/utils/binaryfile.py:1274, in CellBudgetFile._build_index(self)
   1269     self.iposarray.append(
   1270         ipos
   1271     )  # store the position right after header2
   1273     # skip over the data to the next record and set ipos
-> 1274     self._skip_record(header)
   1275     ipos = self.file.tell()
   1277 # convert to numpy arrays

File ~/miniconda3/envs/py39/lib/python3.9/site-packages/flopy/utils/binaryfile.py:1338, in CellBudgetFile._skip_record(self, header)
   1332     nbytes = nlist * (
   1333         np.int32(1).nbytes * 2
   1334         + self.realtype(1).nbytes
   1335         + naux * self.realtype(1).nbytes
   1336     )
   1337 else:
-> 1338     raise Exception(f"invalid method code {imeth}")
   1339 if nbytes != 0:
   1340     self.file.seek(nbytes, 1)

Exception: invalid method code 1085209134`

**Comment by christianlangevin** [comment]
Maybe try this:

```
hds = bf.HeadFile('MYPATH/workspace_2/mf6_2l_ss_rch2.hds', precision='double')
```

**Reply by StellaLiu-gogogo** [reply]
Thanks for the reply. But the head file can be read good. only the .cbb file can't be read. 
any thought?  
thx.

**Reply by jlarsen-usgs** [reply, marked as answer]
Similar to the previous suggestion, maybe try this:
```python
bud = bf.CellBudgetFile(path, precision="double")
```

**Reply by StellaLiu-gogogo** [reply]
thx it works!

**Reply by jlarsen-usgs** [reply]
No problem. The "auto" precision setting has a couple of gotcha's we haven't been able to solve yet for `CellBudgetFile`.

Also, of note. You can access the budget object from your model object after you run a simulation. Here's some pseudocode to make it even easier.

```python
sim = <my flopy simulation object>
m = sim.get_model(<my model name>)
bud = m.output.budget()
```

This method enforces the double precision setting natively and as an additional plus, you don't have to deal with file paths. In my personal opinion, it's the easiest and most robust way to get output from MF6 simulations.

**Reply by StellaLiu-gogogo** [reply]
THX! that is a super helpful suggestion.   However I am indeed curious to know what is the reason it results different precision of my hds file and cbb file. 
How come the hds file was with single precision while cbb file was in double precision? 
Anything related to the datatype of the input? 

I'd like to know how to avoid this issue and always keep the result files in either double of single precision.
